### ISSUE TITLE
Run the setup wizard step by step

### ISSUE THREAD

**Issue by paul121** [farmOS/farmOS#250]
The farmOS setup wizard should have an option to be run as a step-by-step process. This should be the default action presented to users after a new farmOS install. This setup process could also be recommended to users after new features have been added to farmOS, as a way of introducing them to new features & allowing them to be enabled or configured.

Some considerations:
- To start, the setup process will likely just walk through the setup menu tabs:
![Screenshot-2020-2-20 farmOS Setup setup_wizard](https://user-images.githubusercontent.com/3116995/74996706-27f49280-5409-11ea-80fd-034ae4260438.png)
- Because additional tabs can be provided by other modules, it's important that they be included in the step-by-step process
  - Modules will provide a "weight" to order these tabs, which can likely be used to walk through steps
  - The current step may need to be handled by some kind of a hidden `form` element, as part of `next` and `previous` buttons that are included with each setup page.
- In the future, the setup process could be displayed in an improved "modal" UI.
- Additional steps could be displayed if the user is setting up for the first time, eg "Create your first area", "create an asset", etc.
- Other "steps" may be shown for displaying new features to the user.

**Comment by paul121** [comment]
Some original ideas for this are included in this drupal issue: https://www.drupal.org/project/farm/issues/3066502

@mstenta mentioned that the Discourse setup wizard is very nice & could be used as a model to work from:

> Discourse's Setup Wizard is described here: https://meta.discourse.org/t/how-to-use-the-setup-wizard-to-update-your-...
>
> And they have a thread detailing the design decisions they made: https://meta.discourse.org/t/designing-the-first-time-setup-wizard/49512

**Comment by paul121** [comment]
Just finished a first pass on the "stepper" piece of this. A few quick points...

- Adds an additional `control` fieldset to the bottom of each form with button(s):
  - If the form was started from the "Start" page, it will act like a "stepper" with `next` and `back`
  - If an individual tab is clicked, a `submit` button is displayed. For use cases where someone might just want to save one setting.

- Questions:
  - Is a separate `control` fieldset the best way to do this? Better way to visually integrate these buttons? (The `systems_settings_form` doesn't work well with this, but that can easily be fixed)
  - Add simple "progress" at the bottom?

![wizard_demo](https://user-images.githubusercontent.com/3116995/75202469-318b3c80-5720-11ea-988d-340d06b184a8.gif)

**Comment by mstenta** [comment]
Looks great!

> If the form was started from the "Start" page, it will act like a "stepper" with `next` and `back`
> If an individual tab is clicked, a submit button is displayed. For use cases where someone might just want to save one setting.

Curious how this works.

> Is a separate `control` fieldset the best way to do this?

I like it!

> (The systems_settings_form doesn't work well with this, but that can easily be fixed)

Also curious how the form submissions in general will work in this.

> Add simple "progress" at the bottom?

That would be great! Bootstrap provides a component we can use: https://getbootstrap.com/docs/3.4/components/#progress

**Comment by paul121** [comment]
@mstenta and I discussed this in chat. We decided that to properly implement a true setup wizard stepper, it would be better on it's own separate page. Incorporating the tabs with the stepper wizard make it a bit more confusing what is expected from the user. I've implemented this in https://github.com/farmOS/farmOS/pull/251/commits/27e828e6378a8ae3f8545bf6a82a3cc8c1d72ff9.

Designing the setup module in this ways gives the added benefit that setup_forms can be configured to display in either the `wizard`, `setup pages` or both. This will make it easier to add setup steps in the future that only make sense in the true stepper `wizard` on first run. Implemented this in https://github.com/farmOS/farmOS/pull/251/commits/fe445d7e40e0246e5ff7e4586d75581bfdd67b08.

**Comment by paul121** [comment]
Added a bootstrap stepper in https://github.com/farmOS/farmOS/pull/251/commits/77d41bfd0c0d7db07a3cce4f2836c8fdb9a76ae2

![Screenshot-2020-2-27 farmOS Setup Wizard setup_wizard(2)](https://user-images.githubusercontent.com/3116995/75492867-81127800-596d-11ea-96e6-ca58a654d51e.png)

**Comment by paul121** [comment]
Hmm @mstenta I'm curious how we'll want to display the stepper after a new install.

I tried doing it in `hook_init()` as we chatted about... but ran into problems during the installation process (it would redirect to `/farm/setup/wizard` before installation was finished!)

So I tried making a few other checks that would solve the problem (in this commit https://github.com/farmOS/farmOS/pull/251/commits/e7d117cfeb4211dab3f96fb87552e49a1f4d1113) but it still redirects during installation when submitting the "Configure farmOS" form (selecting modules & giving map API keys). FWIW at this stage of the installation process I can open a private browser, go to `localhost/install.php` and continue the installation no problem. So the admin user gets logged in during the install process?

Otherwise...this works great! I think its good to try and only redirect if the user is loading the dashboard so we don't break API requests. But I'm curious if using `hook_init()` might be breaking other things I don't know about.

**Comment by paul121** [comment]
Okay, I found the variable `install_task` that we can use for this. Once the installation is complete, that variable is set to `done`. I've implemented this here https://github.com/farmOS/farmOS/pull/251/commits/599dfc097cdfc2d119eb152d02d02097595c6dcb

Also added a check to make sure the user has `access farm setup` permissions as @mstenta mentioned.

I just tested this on an install - the first thing I saw after installing was the setup wizard! This should be good to go! :+1:
